### PR TITLE
auto-10161: implement zksync automation registry with new forwarder

### PR DESCRIPTION
**Pull request by FelixFan1992**

### CODE REVIEW

File: contracts/src/v0.8/automation/v2_2_zksync/ZKSyncAutomationRegistry2_2.sol
Comment (FelixFan1992) [inline contracts/src/v0.8/automation/v2_2_zksync/ZKSyncAutomationRegistry2_2.sol:1]: this entire directory `v2_2_zksync` will be removed from this PR eventually.

File: contracts/src/v0.8/automation/testhelpers/ZKSyncHashTester.sol
Comment (FelixFan1992) [inline contracts/src/v0.8/automation/testhelpers/ZKSyncHashTester.sol:1]: for testing, will remove eventually

File: contracts/src/v0.8/automation/testhelpers/ZKSyncStoreTester.sol
Comment (FelixFan1992) [inline contracts/src/v0.8/automation/testhelpers/ZKSyncStoreTester.sol:1]: for testing, will remove eventually

File: contracts/src/v0.8/automation/v2_3_zksync/ZKSyncAutomationRegistryLogicB2_3.sol [hunk 1]
```diff
@@ -0,0 +1,454 @@
...
+  ) external returns (bool success, uint256 gasUsed) {
+    _preventExecution();
+
+    if (s_hotVars.paused) revert RegistryPaused();
+    Upkeep memory upkeep = s_upkeep[id];
+
+    uint256 l1GasUsed;
+    (success, gasUsed, l1GasUsed) = _performUpkeep(upkeep.forwarder, upkeep.performGas, performData);
+
+    if (upkeep.performGas < l1GasUsed + gasUsed) {
```
Comment (FelixFan1992) [inline contracts/src/v0.8/automation/v2_3_zksync/ZKSyncAutomationRegistryLogicB2_3.sol:234]: make sure the total gas used is bound by upkeep gas limit

File: contracts/src/v0.8/automation/v2_3_zksync/ZKSyncAutomationRegistryBase2_3.sol [hunk 2]
```diff
@@ -0,0 +1,1226 @@
...
+  uint256 internal constant PPB_BASE = 1_000_000_000;
+  uint32 internal constant UINT32_MAX = type(uint32).max;
+  // The first byte of the mask can be 0, because we only ever have 31 oracles
+  uint256 internal constant ORACLE_MASK = 0x0001010101010101010101010101010101010101010101010101010101010101;
+  uint8 internal constant UPKEEP_VERSION_BASE = 4;
+
+  // Next block of constants are only used in maxPayment estimation during checkUpkeep simulation
+  // These values are calibrated using hardhat tests which simulate various cases and verify that
+  // the variables result in accurate estimation
+  uint256 internal constant REGISTRY_CONDITIONAL_OVERHEAD = 92_000; // Fixed gas overhead for conditional upkeeps
```
Comment (FelixFan1992) [inline contracts/src/v0.8/automation/v2_3_zksync/ZKSyncAutomationRegistryBase2_3.sol:47]: conditional and log trigger overheads are reduced slightly due to simpler logic in hot path perform

File: contracts/src/v0.8/automation/v2_3_zksync/ZKSyncAutomationRegistryBase2_3.sol [hunk 3]
```diff
@@ -0,0 +1,1226 @@
...
+   * @member l1GasUsed the L1 gas used by this upkeep in perform
+   * @member dedupID unique ID used to dedup an upkeep/trigger combo
+   */
+  struct UpkeepTransmitInfo {
+    Upkeep upkeep;
+    bool earlyChecksPassed;
+    bool performSuccess;
+    Trigger triggerType;
+    uint256 gasUsed;
+    uint256 l1GasUsed;
```
Comment (FelixFan1992) [inline contracts/src/v0.8/automation/v2_3_zksync/ZKSyncAutomationRegistryBase2_3.sol:324]: removed calldata weight
added l1GasUsed

File: contracts/src/v0.8/automation/v2_3_zksync/ZKSyncAutomationRegistryBase2_3.sol [hunk 4]
```diff
@@ -0,0 +1,1226 @@
...
+  );
+  event UpkeepCharged(uint256 indexed id, PaymentReceipt receipt);
+  event UpkeepPrivilegeConfigSet(uint256 indexed id, bytes privilegeConfig);
+  event UpkeepReceived(uint256 indexed id, uint256 startingBalance, address importedFrom);
+  event UpkeepRegistered(uint256 indexed id, uint32 performGas, address admin);
+  event UpkeepTriggerConfigSet(uint256 indexed id, bytes triggerConfig);
+  event UpkeepUnpaused(uint256 indexed id);
+  event Unpaused(address account);
+
+  // for testing, remove later
```
Comment (FelixFan1992) [inline contracts/src/v0.8/automation/v2_3_zksync/ZKSyncAutomationRegistryBase2_3.sol:501]: for testing purpose, will remove eventually

File: contracts/src/v0.8/automation/ZKSyncAutomationForwarder.sol [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+  // @notice the user's target contract address
+  address private immutable i_target;
+
+  // @notice the shared logic address
+  address private immutable i_logic;
+
+  IAutomationRegistryConsumer private s_registry;
+
+  // for testing purpose, remove later
```
Comment (FelixFan1992) [inline contracts/src/v0.8/automation/ZKSyncAutomationForwarder.sol:26]: for testing purpose, remove later

File: contracts/src/v0.8/automation/ZKSyncAutomationForwarder.sol [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+  function forward(
+    uint256 gasAmount,
+    bytes memory data
+  ) external returns (bool success, uint256 gasUsed, uint256 l1GasUsed) {
+    if (msg.sender != address(s_registry)) revert InvalidCaller(msg.sender);
+    address target = i_target;
+
+    uint256 g1 = gasleft();
+    bytes memory returnData;
+    (success, returnData) = GAS_BOUND_CALLER.delegatecall{gas: gasAmount}(
```
Comment (FelixFan1992) [inline contracts/src/v0.8/automation/ZKSyncAutomationForwarder.sol:56]: use delegatecall to preserve the `msg.sender`

File: contracts/src/v0.8/automation/interfaces/zksync/IGasBoundCaller.sol [hunk 7]
```diff
@@ -0,0 +1,8 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity 0.8.19;
+
+address constant GAS_BOUND_CALLER = address(0xc706EC7dfA5D4Dc87f29f859094165E8290530f5);
```
Comment (shileiwill) [inline contracts/src/v0.8/automation/interfaces/zksync/IGasBoundCaller.sol:4]: I guess this contract is about [this](https://docs.zksync.io/zk-stack/components/smart-contracts/system-contracts#gasboundcaller). which network is this address deployed to?

File: contracts/src/v0.8/automation/v2_2_zksync/ZKSyncAutomationRegistry2_2.sol [hunk 8]
```diff
@@ -0,0 +1,423 @@
...
+import {Chainable} from "../Chainable.sol";
+import {IERC677Receiver} from "../../shared/interfaces/IERC677Receiver.sol";
+import {OCR2Abstract} from "../../shared/ocr2/OCR2Abstract.sol";
+import {SYSTEM_CONTEXT_CONTRACT} from "../interfaces/zksync/ISystemContext.sol";
+
+/**
+ * @notice Registry for adding work for Chainlink nodes to perform on client
+ * contracts. Clients must support the AutomationCompatibleInterface interface.
+ */
+contract ZKSyncAutomationRegistry2_2 is ZKSyncAutomationRegistryBase2_2, OCR2Abstract, Chainable, IERC677Receiver {
```
Comment (shileiwill) [inline contracts/src/v0.8/automation/v2_2_zksync/ZKSyncAutomationRegistry2_2.sol:17]: if we have zksync_v23, do we still need zksync_v22?

File: contracts/src/v0.8/automation/v2_3_zksync/ZKSyncAutomationRegistryLogicC2_3.sol [hunk 9]
```diff
@@ -0,0 +1,632 @@
...
+
+  /**
+   * @dev see AutomationRegistry master contract for constructor description
+   */
+  constructor(
+    address link,
+    address linkUSDFeed,
+    address nativeUSDFeed,
+    address fastGasFeed,
+    address automationForwarderLogic,
```
Comment (FelixFan1992) [inline contracts/src/v0.8/automation/v2_3_zksync/ZKSyncAutomationRegistryLogicC2_3.sol:25]: no `allowedReadOnlyAddress`

File: contracts/src/v0.8/automation/v2_3_zksync/ZKSyncAutomationRegistryLogicC2_3.sol [hunk 10]
```diff
@@ -0,0 +1,632 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity 0.8.19;
+
+import {ZKSyncAutomationRegistryBase2_3} from "./ZKSyncAutomationRegistryBase2_3.sol";
+import {EnumerableSet} from "../../vendor/openzeppelin-solidity/v4.7.3/contracts/utils/structs/EnumerableSet.sol";
+import {Address} from "../../vendor/openzeppelin-solidity/v4.7.3/contracts/utils/Address.sol";
+import {IZKSyncAutomationForwarder} from "../interfaces/zksync/IZKSyncAutomationForwarder.sol";
```
Comment (FelixFan1992) [inline contracts/src/v0.8/automation/v2_3_zksync/ZKSyncAutomationRegistryLogicC2_3.sol:7]: different interface

File: contracts/src/v0.8/automation/v2_3_zksync/ZKSyncAutomationRegistryLogicC2_3.sol [hunk 11]
```diff
@@ -0,0 +1,632 @@
...
+  constructor(
+    address link,
+    address linkUSDFeed,
+    address nativeUSDFeed,
+    address fastGasFeed,
+    address automationForwarderLogic,
+    PayoutMode payoutMode,
+    address wrappedNativeTokenAddress
+  )
+    ZKSyncAutomationRegistryBase2_3(
```
Comment (FelixFan1992) [inline contracts/src/v0.8/automation/v2_3_zksync/ZKSyncAutomationRegistryLogicC2_3.sol:29]: zksync version of base

File: contracts/src/v0.8/automation/v2_3_zksync/ZKSyncAutomationRegistryLogicC2_3.sol [hunk 12]
```diff
@@ -0,0 +1,632 @@
...
+   * @notice returns the admin's privilege config
+   */
+  function getAdminPrivilegeConfig(address admin) external view returns (bytes memory) {
+    return s_adminPrivilegeConfig[admin];
+  }
+
+  /**
+   * @notice returns the upkeep's forwarder contract
+   */
+  function getForwarder(uint256 upkeepID) external view returns (IZKSyncAutomationForwarder) {
```
Comment (FelixFan1992) [inline contracts/src/v0.8/automation/v2_3_zksync/ZKSyncAutomationRegistryLogicC2_3.sol:562]: different forwarder interface

File: contracts/src/v0.8/automation/v2_3_zksync/ZKSyncAutomationRegistryLogicB2_3.sol [hunk 13]
```diff
@@ -0,0 +1,454 @@
...
+import {ZKSyncAutomationRegistryBase2_3} from "./ZKSyncAutomationRegistryBase2_3.sol";
+import {EnumerableSet} from "../../vendor/openzeppelin-solidity/v4.7.3/contracts/utils/structs/EnumerableSet.sol";
+import {Address} from "../../vendor/openzeppelin-solidity/v4.7.3/contracts/utils/Address.sol";
+import {ZKSyncAutomationRegistryLogicC2_3} from "./ZKSyncAutomationRegistryLogicC2_3.sol";
+import {Chainable} from "../Chainable.sol";
+import {IERC20Metadata as IERC20} from "../../vendor/openzeppelin-solidity/v4.8.3/contracts/token/ERC20/extensions/IERC20Metadata.sol";
+import {SafeERC20} from "../../vendor/openzeppelin-solidity/v4.8.3/contracts/token/ERC20/utils/SafeERC20.sol";
+import {SafeCast} from "../../vendor/openzeppelin-solidity/v4.8.3/contracts/utils/math/SafeCast.sol";
+
+contract ZKSyncAutomationRegistryLogicB2_3 is ZKSyncAutomationRegistryBase2_3, Chainable {
```
Comment (FelixFan1992) [inline contracts/src/v0.8/automation/v2_3_zksync/ZKSyncAutomationRegistryLogicB2_3.sol:13]: no `allowedReadOnlyAddress`
use zksync version of base

File: contracts/src/v0.8/automation/v2_3_zksync/ZKSyncAutomationRegistryLogicA2_3.sol [hunk 14]
```diff
@@ -0,0 +1,282 @@
...
+import {UpkeepTranscoderInterfaceV2} from "../interfaces/UpkeepTranscoderInterfaceV2.sol";
+import {MigratableKeeperRegistryInterfaceV2} from "../interfaces/MigratableKeeperRegistryInterfaceV2.sol";
+import {IERC20Metadata as IERC20} from "../../vendor/openzeppelin-solidity/v4.8.3/contracts/token/ERC20/extensions/IERC20Metadata.sol";
+import {SafeERC20} from "../../vendor/openzeppelin-solidity/v4.8.3/contracts/token/ERC20/utils/SafeERC20.sol";
+import {IERC677Receiver} from "../../shared/interfaces/IERC677Receiver.sol";
+
+/**
+ * @notice Logic contract, works in tandem with AutomationRegistry as a proxy
+ */
+contract ZKSyncAutomationRegistryLogicA2_3 is ZKSyncAutomationRegistryBase2_3, Chainable, IERC677Receiver {
```
Comment (FelixFan1992) [inline contracts/src/v0.8/automation/v2_3_zksync/ZKSyncAutomationRegistryLogicA2_3.sol:21]: use zksync version of base
use zksync version of forwarder and forwarder interface
no `allowedReadOnlyAddress`

File: contracts/src/v0.8/automation/v2_3_zksync/ZKSyncAutomationRegistryLogicC2_3.sol (hunk 9 above)
Comment (FelixFan1992) [reply contracts/src/v0.8/automation/v2_3_zksync/ZKSyncAutomationRegistryLogicC2_3.sol:25]: remove this to minimize changes

File: contracts/src/v0.8/automation/interfaces/zksync/IZKSyncAutomationForwarder.sol [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+pragma solidity ^0.8.0;
+
+import {ITypeAndVersion} from "../../../shared/interfaces/ITypeAndVersion.sol";
+import {IAutomationRegistryConsumer} from "./../IAutomationRegistryConsumer.sol";
+
+interface IZKSyncAutomationForwarder is ITypeAndVersion {
+  function forward(
+    uint256 gasAmount,
+    bytes memory data
+  ) external returns (bool success, uint256 gasUsed, uint256 l1GasUsed);
```
Comment (FelixFan1992) [inline contracts/src/v0.8/automation/interfaces/zksync/IZKSyncAutomationForwarder.sol:11]: investigate if we can remove `l1GasUsed`

File: contracts/src/v0.8/automation/v2_3_zksync/ZKSyncAutomationRegistryLogicB2_3.sol (hunk 1 above)
Comment (FelixFan1992) [reply contracts/src/v0.8/automation/v2_3_zksync/ZKSyncAutomationRegistryLogicB2_3.sol:234]: can we move this detail into forwarder?

File: contracts/src/v0.8/automation/ZKSyncAutomationForwarder.sol [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+
+    uint256 g1 = gasleft();
+    bytes memory returnData;
+    // solhint-disable-next-line avoid-low-level-calls
+    (success, returnData) = GAS_BOUND_CALLER.delegatecall{gas: gasAmount}(
+      abi.encodeWithSelector(IGasBoundCaller.gasBoundCall.selector, target, gasAmount, data)
+    );
+    uint256 pubdataGasSpent;
+    if (success) {
+      (, pubdataGasSpent) = abi.decode(returnData, (bytes, uint256));
```
Comment (FelixFan1992) [inline contracts/src/v0.8/automation/ZKSyncAutomationForwarder.sol:61]: explore to combine `pubdataGasSpent` with `gasUsed`

File: contracts/src/v0.8/automation/v2_3_zksync/ZKSyncAutomationRegistry2_3.sol [hunk 17]
```diff
@@ -0,0 +1,392 @@
...
+      _updateTriggerMarker(report.upkeepIds[i], blocknumber, upkeepTransmitInfo[i]);
+    }
+    // No upkeeps to be performed in this report
+    if (transmitVars.numUpkeepsPassedChecks == 0) {
+      return;
+    }
+
+    // This is the overall gas overhead that will be split across performed upkeeps
+    // Take upper bound of 16 gas per callData bytes
+    gasOverhead = gasOverhead + (16 * msg.data.length) + ACCOUNTING_FIXED_GAS_OVERHEAD - gasleft();
```
Comment (FelixFan1992) [inline contracts/src/v0.8/automation/v2_3_zksync/ZKSyncAutomationRegistry2_3.sol:180]: possible underflow

File: contracts/src/v0.8/automation/v2_3_zksync/ZKSyncAutomationRegistry2_3.sol [hunk 18]
```diff
@@ -0,0 +1,392 @@
...
+      uint256 nativeUSD = _getNativeUSD(hotVars);
+      for (uint256 i = 0; i < report.upkeepIds.length; i++) {
+        if (upkeepTransmitInfo[i].earlyChecksPassed) {
+          if (i == 0 || upkeepTransmitInfo[i].upkeep.billingToken != upkeepTransmitInfo[i - 1].upkeep.billingToken) {
+            billingTokenParams = _getBillingTokenPaymentParams(hotVars, upkeepTransmitInfo[i].upkeep.billingToken);
+          }
+          PaymentReceipt memory receipt = _handlePayment(
+            hotVars,
+            PaymentParams({
+              gasLimit: upkeepTransmitInfo[i].gasUsed,
```
Comment (FelixFan1992) [inline contracts/src/v0.8/automation/v2_3_zksync/ZKSyncAutomationRegistry2_3.sol:194]: option 1: fixed gas overhead + `upkeepTransmitInfo[i].gasUsed + upkeepTransmitInfo[i].l1GasUsed`

File: contracts/src/v0.8/automation/chains/ZKSyncModule.sol [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+pragma solidity 0.8.19;
+
+import {ChainModuleBase} from "./ChainModuleBase.sol";
+import {SYSTEM_CONTEXT_CONTRACT} from "../interfaces/zksync/ISystemContext.sol";
+
+contract ZKSyncModule is ChainModuleBase {
+  uint256 private constant FIXED_GAS_OVERHEAD = 5_000;
+
+  function getMaxL1Fee(uint256 maxCalldataSize) external view override returns (uint256) {
+    return maxCalldataSize * SYSTEM_CONTEXT_CONTRACT.gasPerPubdataByte() * SYSTEM_CONTEXT_CONTRACT.gasPrice();
```
Comment (FelixFan1992) [inline contracts/src/v0.8/automation/chains/ZKSyncModule.sol:11]: update this

**Pull request closed without merging**